[WIP] soundfile plugin architecture

danomatika opened the pull request:
This PR is a feature branch for #855 which abstracts the file type specific function via a `t_soundfile_filetype` struct and function pointers. New file types can be implemented and added via `soundfile_addfiletype()`.

My initial approach is to just take the pragmatic refactoring I already did and abstract it. For a more *proper* architecture, I think the implementations should have more direct control over sample conversion and opening/closing files. We can look at this for "round two."

danomatika commented:
Some things I'm thinking about:

Currently, the `sf_filetypes` array is created with a limited size. It might be better to malloc and use an array of function pointers.

The writesf~ `-nextstep` flag is currently broken as it is superseded by `-next`. I'd like to provide `-nextstep` as an alias, but I didn't come up with a clean way to do this yet.

The `t_sf_*fn` naming is a bit awkward.

Should the file type structs / pointers be const in the global array?

Spacechild1 commented:
This is nice! I'm terribly busy right now, but I'll have a look and see what the API would need for compressed audio codecs, e.g. MP3 via liblame. Unless you're faster :-)

danomatika commented:
This is something along the lines of what I'm thinking now:

~~~
typedef struct _soundfile
{
    void *sf_file;                /**< implementation-specific file descriptor */
    t_soundfile_filetype *sf_filetype; /**< file type */
    t_soundfile_info sf_info; /**< format info */
    void *sf_data;            /**< implementation-specific data */
} t_soundfile;

void soundfile_clear(t_soundfile *sf);

void soundfile_copy(t_soundfile *dst, const t_soundfile *src);
~~~

This could be the main soundfile type passed between functions. The `t_soundfile_filetype` should also have implementation-specific open and close functions as we cannot assume they would all use a simple int fd:

~~~
typedef int (*t_sf_openfn)(const char *filepath, t_soundfile *sf);

typedef int (*t_sf_closefn)(t_soundfile *sf);
~~~

The `t_sf_openfn` could alternately take a `const char *filename, const char *dir` arguments instead of the combined path. I need to find a way to hook into Pd's `open_via_path()` and `canvas_open()` functions to get the file path and parent dir instead of *also* opening the fd.

Additionally, the current `t_soundfile_filetype` struct should probably be renamed `t_soundfile_backend` as there could be implementations that handle multiple types.

Spacechild1 commented:
Looks good. The questions is how to control codec dependant options with Pd messages in a generic way.

danomatika commented:
I was thinking more of a generic argc argv approach with possibly and outlet as well for meta message output. These could be processed when the implementation-specific flag is passed ala “-sndfile arg arg” or maybe via a second inlet.

My first use case is reading/writing instrument chunk data like sample points and tuning info. One problem with this is timing but maybe that should be left up to the user as in writesf~ open creates the file and header, then you write meta data chunks, then you write sound data, then close. It would be more complicated to implicitly “remember” metadata winces Pds current approach is to recreate the file each time it is written to without reading existing chunk data.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 13, 2020, at 5:56 PM, Christof Ressi <notifications@github.com> wrote:
> 
> Looks good. The questions is how to control codec dependant options with Pd messages in a generic way.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

danomatika commented:
I think now the best way around the file descriptor issue is to simply use `open_via_path()` and `canvas_open` to get the filename and parent directory in order to build a path to hand over to the file type backends. So far, I'm refactoring things around this but one downside is that we call `open()` to get the path, immediately `close()`, then basically open again inside the implementation-specific code. As we don't shared the file descriptor or a basic header check buffer, each implementation now opens the file and inspects the header to determine if it's something it can work with. If not, it closes and we check the next file type backend. My only concern is if this if this is expensive... or not.

Spacechild1 commented:
> So far, I'm refactoring things around this but one downside is that we call open() to get the path, immediately close(), then basically open again inside the implementation-specific code. 

Yes, but we already do this anyway in `canvas_open()`: if the path is absolute, we don't check and assume it is correct, but for relative paths, we iterate over the canvas path list - which can lead to several calls to `sys_trytoopenone` before we can finally resolve the path. `sys_trytoopenone` contains at least an `open()` call - plus a `fstat()` call on Linux! - and a `close()` call on failure. For larger soundfiles, the actual file IO will easily outweight the overhead of another `open()/close()` pair, so we probably shouldn't worry too much about it.

I made a small benchmark program which gives a rough idea about the overhead:
[stat-test.zip](https://github.com/pure-data/pure-data/files/4206960/stat-test.zip)
This is what I get on my system:
```
enter file name: C:/Software/nan-test.exe
enter number of iterations: 10
benchmarking stat():
count: 10, time (ms): 0.917
benchmarking open():
count: 10, time (ms): 0.189
reading file:
read 326712 bytes, time (ms): 0.204
```
On my system, 10 pairs of  `open()/close()` take the same time as reading a 326 kB file into memory. Note that `stat()` is roughly 5 time slower than `open()/close()` (`sys_trytoopenone` uses `fstat()`, though).

PS: I wrote it on Windows, so maybe I forgot a header for Linux/macOS.

danomatika commented:
Yeah, I had looked through s_path.c and saw it basically tried to poke paths i til it succeeds, so I figured it wasn’t a big deal. It’s still good to question assumptions, though, and this simplified the changes needed. Hopefully I can get to this today, if there is a midday nap now...

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 14, 2020, at 10:53 PM, Christof Ressi <notifications@github.com> wrote:
> 
> So far, I'm refactoring things around this but one downside is that we call open() to get the path, immediately close(), then basically open again inside the implementation-specific code.
> 
> Yes, but we already do this anyway in canvas_open(): if the path is absolute, we don't check and assume it is correct, but for relative paths, we iterate over the canvas path list - which can lead to several calls to sys_trytoopenone before we can finally resolve the path. sys_trytoopenone contains at least an open() call - plus a fstat() call on Linux! - and a close() call on failure. For larger soundfiles, the actual file IO will easily outweight the overhead of another open()/close() pair, so we probably shouldn't worry too much about it.
> 
> I made a small benchmark program which gives a rough idea about the overhead:
> stat-test.zip
> This is what I get on my system:
> 
> enter file name: C:/Software/nan-test.exe
> enter number of iterations: 10
> benchmarking stat():
> count: 10, time (ms): 0.917
> benchmarking open():
> count: 10, time (ms): 0.189
> reading file:
> read 326712 bytes, time (ms): 0.204
> On my system, 10 pairs of open()/close() take the same time as reading a 326 kB file into memory. Note that stat() is roughly 5 time slower than open()/close() (sys_trytoopenone uses fstat(), though).
> 
> PS: I wrote it on Windows, so maybe I forgot a header for Linux/macOS.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Spacechild1 commented:
> It’s still good to question assumptions, though,

Of course! I had no idea about the actual cost of `open()/close()`, so benchmarking it was interesting.

umlaeute commented:
i *think* your benchmark might not measure the thing you intended to measure.

 in the end, `open()` wraps a system call and it's performance will depend on a lot of external factors, like the filesystem in use and the speed of the floppy drive.

the more calls to `open()` there are, the higher the chance to get a drop-out.
this is why i'm not sure whether we should drop the header-caching optimisation (Pd attempts to read the header once, so the backends can do a quick check whether they will *not* be able to read that specific file without having to do `open()` that same file again).

of course, the OS *might* cache the file, so subsequent calls to `open()` will be not that bad, but nobody knows (except for the OS itself)

umlaeute commented:
> UPDATE: The main need for this is that we cannot assume external backends work with low-level file descriptors, ex. libsdnfile uses an SFFILE struct modeled after C's FILE.

actually, libsndfile provides `sf_open_fd()` to open a file given it's file-descriptor, so that should not be a problem.

> the actual file IO will easily outweight the overhead of another open()/close() pair, so we probably shouldn't worry too much about it.

ah well, but the actual I/O might be done in another thread, whereas the calls to `open()` will happen in the main thread.

Spacechild1 commented:
> ah well, but the actual I/O might be done in another thread, whereas the calls to open() will happen in the main thread.

Actually, all the calls to `open()` could/should also handed over to another thread in the future. I have always wondered, why the canvas path list is not simply protected by a reader-writer-lock. Since the list is rarely modified, but accessed a lot, this should work fine. What am I missing?

danomatika commented:
I looked at some libmp3lame examples and it doesn't have any built-in file handling. Since more of these kinds of libraries are probably written in C, I imagine they probably work with file descriptors and/or FILE. I think I will hold off on the implementation-specific open/close and stay with Pd's reuse of the fd. I also note that `soundfiler_read()` is reusing the fd again via `fdopen()` for double checking the number of frames available.

danomatika commented:
I've consolidated `t_soundfile_info`, the file descriptor, and filetype into a is single `t_soundfile` struct and will push these changes soon.

Any suggested for renaming `t_soundfile_filetype`? I was originally thinking `t_soundfile_backend` but perhaps simply `t_soundfile_type` would work as it simultaneously refers to both _file_ type as well as _backend_ type.

From a user's perspective, the "backend" is both the file extension *and* the explicit type flags, ie. `-aiff` or `-wav`. My intention is that additional implementations would build onto this so a libsndfile "backend" would be implicitly chosen if a user passes a file type which is not loaded by the built-in backends and explicitly chosen if the backs type's flag is used, ie. `-sndfile`.

danomatika commented:
The latest commit brings the `t_soundfile` consolidation as well as filetype backend open, close, resamples., & writesamples functions. The latter 2 use byte buffers and I'm thinking of changing them to use sample frames instead.

Also, I'm thinking now that `t_soundfile.sf_bytelimit` should be in sample frames instead of bytes, ie. `sf_framelimit` as a data size in bytes is not as useful for sound file compressed format sound files.

danomatika commented:
Latest commits add filetype backend flags to `readsf_open()` so `[readsf~], `[writesf~], and `[soundfiler~]` now take explicit backend overrides. This should allow choosing custom or 3rd party backend instead auto-detection.

Also, `soundfiler_read()` was always reading through the file samples to find the number of frames, irregardless if a tablename was supplied.  Since the filetype backends now read the audio data size from the file headers (and or calculate based on header size and file size), I think it makes more sense to use that instead of reading. If the calculated size is max size or raw head info is being used, then the samples are read.

danomatika commented:
Latest commits:

* Fix a couple bugs in the `[readsf~]` thread communication
* Use lower-case filetype error naming, ie; "aiff 32 bit int not supported" to match "-aiff"
* If the filetype was set after a read fail, it is included in the error print ie. `[open -next file.wav<` -> "readsf: file.wav: unknown or bad header (next)". This is to help make the distinction between "I couldn't find anything that could read this file" and "you asked me to read a .wav as a next file and I couldn't understand it."
* Added basic filetype-specific meta data read/write plumbing:
  - read: Use "-meta" flag with `[soundfiler]` open message, thus optionally passes sound filer's 2nd "info" outlet to the filetype readheaderfn so implementations can pass out meta chunk data as lists, etc.
  - write: Use `[writesf~]` "meta" message after calling "open" to pass meta data as lists. The idea is that the file's header can be written to and updated *before* writing sound data.

danomatika commented:
I think this could use some testing but I feel that most everything is there, so far. @umlaeute  @Spacechild1 @Lucarda 

I would also like some feedback on the `t_soundfile_filetype` renaming ideas.

danomatika commented:
Note: In the end there was no need for filetype-specific sample conversion to/frame the t_sample buffers as soundfile read/write was already doing this. The filetype readsamplesfn & writesamplesfn basically get int or float data ready to go in raw buffers.

Lucarda commented:
>I think this could use some testing but I feel that most everything is there, so far



when i finish the recordings (or 1 only) and close the test-patch(1) Pd Crashes (on windows64).
Didn't happend in my previous build of `update/soundfile`.



(1)[soundfile-test-v7.zip](https://github.com/pure-data/pure-data/files/4220385/soundfile-test-v7.zip)



Some msys2 GDB:

---------------------------------------------------------
```
$ gdb --args pd

Type "apropos word" to search for commands related to "word"...
Traceback (most recent call last):
  File "<string>", line 3, in <module>
ModuleNotFoundError: No module named 'libstdcxx'
/etc/gdbinit:6: Error in sourced command file:
Error while executing Python code.
Reading symbols from pd...(no debugging symbols found)...done.


(gdb) run


Starting program: /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd
[New Thread 4848.0x135c]
warning: Application "\??\D:\00-volatil\pure-data-feature-soundfile-backend\pd-0.50.2\bin\wish85.exe" not found in cache
[New Thread 4848.0x9d4]
[New Thread 4848.0x1278]
[New Thread 4848.0x1274]
[New Thread 4848.0x13fc]
[New Thread 4848.0x1318]
[New Thread 4848.0xd74]
[New Thread 4848.0x1294]
[Thread 4848.0x9d4 exited with code 0]
[Thread 4848.0x1278 exited with code 0]

Thread 4 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 4848.0x1274]
0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll

   
   (gdb) bt
   
   
#0  0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll
#1  0x0000000064945092 in pthread_create_wrapper ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/libwinpthread-1.dll
#2  0x00007ffbb13c9967 in sinf () from /c/Windows/system32/msvcrt.dll
#3  0x00007ffbb13c9a0d in msvcrt!_endthreadex ()
   from /c/Windows/system32/msvcrt.dll
#4  0x00007ffbb10a1611 in KERNEL32!BaseThreadInitThunk ()
   from /c/Windows/system32/KERNEL32.DLL
#5  0x00007ffbb38864ad in ntdll!RtlUserThreadStart ()
   from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)


(gdb) bt


#0  0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll
#1  0x0000000064945092 in pthread_create_wrapper ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/libwinpthread-1.dll
#2  0x00007ffbb13c9967 in sinf () from /c/Windows/system32/msvcrt.dll
#3  0x00007ffbb13c9a0d in msvcrt!_endthreadex ()
   from /c/Windows/system32/msvcrt.dll
#4  0x00007ffbb10a1611 in KERNEL32!BaseThreadInitThunk ()
   from /c/Windows/system32/KERNEL32.DLL
#5  0x00007ffbb38864ad in ntdll!RtlUserThreadStart ()
   from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)


(gdb) bt


#0  0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll
#1  0x0000000064945092 in pthread_create_wrapper ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/libwinpthread-1.dll
#2  0x00007ffbb13c9967 in sinf () from /c/Windows/system32/msvcrt.dll
#3  0x00007ffbb13c9a0d in msvcrt!_endthreadex ()
   from /c/Windows/system32/msvcrt.dll
#4  0x00007ffbb10a1611 in KERNEL32!BaseThreadInitThunk ()
   from /c/Windows/system32/KERNEL32.DLL
#5  0x00007ffbb38864ad in ntdll!RtlUserThreadStart ()
   from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)


(gdb) bt


#0  0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll
#1  0x0000000064945092 in pthread_create_wrapper ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/libwinpthread-1.dll
#2  0x00007ffbb13c9967 in sinf () from /c/Windows/system32/msvcrt.dll
#3  0x00007ffbb13c9a0d in msvcrt!_endthreadex ()
   from /c/Windows/system32/msvcrt.dll
#4  0x00007ffbb10a1611 in KERNEL32!BaseThreadInitThunk ()
   from /c/Windows/system32/KERNEL32.DLL
#5  0x00007ffbb38864ad in ntdll!RtlUserThreadStart ()
   from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)


(gdb) bt


#0  0x0000000067363f05 in pd!open_soundfile_via_path ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/pd.dll
#1  0x0000000064945092 in pthread_create_wrapper ()
   from /d/00-volatil/pure-data-feature-soundfile-backend/pd-0.50.2/bin/libwinpthread-1.dll
#2  0x00007ffbb13c9967 in sinf () from /c/Windows/system32/msvcrt.dll
#3  0x00007ffbb13c9a0d in msvcrt!_endthreadex ()
   from /c/Windows/system32/msvcrt.dll
#4  0x00007ffbb10a1611 in KERNEL32!BaseThreadInitThunk ()
   from /c/Windows/system32/KERNEL32.DLL
#5  0x00007ffbb38864ad in ntdll!RtlUserThreadStart ()
   from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb)
```

Lucarda commented:
there might be more BT but i stopped there.

danomatika commented:
I see the crash as well.

Also, if you haven't already, you might be able to get a better backtrace if you build Pd with debugging symbols:

./configure --enable-debug CFLAGS="-g"

> On Feb 18, 2020, at 5:44 PM, Lucas Cordiviola <notifications@github.com> wrote:
> 
> there might be more BT but i stopped there.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pure-data/pure-data/pull/894?email_source=notifications&email_token=AADVK7OQ6IPZYR4ZVIW4SMDRDQGABA5CNFSM4KUFQYK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMCWVMI#issuecomment-587557553>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADVK7MUVATJG4KTTBBALWTRDQGABANCNFSM4KUFQYKQ>.
> 

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

Lucarda commented:
> ./configure --enable-debug CFLAGS="-g"

cool. I always do `make app` and then `make clean`. Will see the next build. I don't have to do the make app because it will get stripped? or it will work with debug symbols included?

danomatika commented:
If the changes are only to rebuild the core `pd.exe` binary, you can simply copy it into the app folder. This is what I do for most testing on macOS: make, copy `src/pd` into a pre-build macOS .app, and start the binary from inside the bundle.

On windows it would probably be something like this:

    make
    cp src/pd.exe pd-0.51-0/bin && ./pd-0.51-0/bin/pd.exe

danomatika commented:
Update: latest commit should fix the thread crash on exit. The sf fd wasn't being initialized correctly.

Lucarda commented:
will test. I might build manually also including `pd.dll`

Spacechild1 commented:
> If the changes are only to rebuild the core pd.exe binary, you can simply copy it into the app folder. 

pd.exe is only a tiny stub, it is pd.dll which contains the actual code. Personally, I simply copy the pd.dll to the app folder on the command line right after make:

`make; cp ./src/pd.dll ./pd-0.50.2/bin`

(assuming that `make app` has already been run at least once)

danomatika commented:
Ok, replace my mentions of ".exe" with .dll" but same idea. On macOS and Linux, there is only a simple executable.

> On Feb 18, 2020, at 6:11 PM, Christof Ressi <notifications@github.com> wrote:
> 
> If the changes are only to rebuild the core pd.exe binary, you can simply copy it into the app folder.
> 
> pd.exe is only a tiny stub, it is pd.dll which contains the actual code. Personally, I simply copy the pd.dll to the app folder on the command line right after make:
> 
> make; cp ./src/pd.dll ./pd-0.50.2/bin
> 
> (assuming that make app has already been run at least once)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/pure-data/pure-data/pull/894?email_source=notifications&email_token=AADVK7NOSSO7DHAILQATCWLRDQJFNA5CNFSM4KUFQYK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMCZ57Y#issuecomment-587570943>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADVK7NYJ3H7BATWLVMXXZLRDQJFNANCNFSM4KUFQYKQ>.
> 

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

Lucarda commented:
All good here with 5aabd68.

The problem is gone and checks for the files with VLC, Soundforge and sndfile-info are Ok.

Lucarda commented:
for the sake of completeness I could never open the .au files on Soundforge. (on current Pd or in this overhaul).

danomatika commented:
Save them as .snd files.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 18, 2020, at 7:06 PM, Lucas Cordiviola <notifications@github.com> wrote:
> 
> for the sake of completeness I could never open the .au files on Soundforge. (on current Pd or in this overhaul).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Lucarda commented:
> Save them as .snd files.

I have already tried that but :( nop.

danomatika commented:
Does it have an error info? sndfile-info, soxi, and VLC read them ok.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 18, 2020, at 7:44 PM, Lucas Cordiviola <notifications@github.com> wrote:
> 
> Save them as .snd files.
> 
> I have already tried that but :( nop.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

danomatika commented:
As an experiment, I started writing meta read/write for AIFF text chunks (name, author, copyright, comment). What I found is:

* A file format with an _undefined_ chunk order means existing chunk bytes may have to be moved to make space for new/updated data. We limit this by only allowing meta writing *after* the file is opened but *before* audio data is written.
* The file descriptor opened in `create_soundfile()` needs to be `O_RDWR` as the chunk order may need to be read from the file when writing metadata, i.e. data chunk position is no longer static, so it has to be found each time. I suppose one way around this is to use the sf->sf_data pointer to hold a temporary struct with the offset positions for everything.

umlaeute commented:
>     *  We limit this by only allowing meta writing _after_ the file is opened but _before_ audio data is written.

is this something that should be enforced on the public API (as in: all backends are forced to this behaviour), or something a backend can decide for itself? that is: we just pass the meta-request to the backend, and it can refuse to accept it (with an error message).

(obviously any error message must `post()`ed in the master-thread (or properly locked))

danomatika commented:
Currently, this is enforced via state checking in writesf_meta() and is applied to all backends. If the backend doesn’t provide a writemetafn, a poets says the becker doesn’t support meta data.

As for posting, the backend implementations use both post() and error() which could happen in a background thread. I suppose this should be changed?

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Feb 19, 2020, at 9:12 AM, umläute <notifications@github.com> wrote:
> 
> *  We limit this by only allowing meta writing _after_ the file is opened but _before_ audio data is written.
> is this something that should be enforced on the public API (as in: all backends are forced to this behaviour), or something a backend can decide for itself? that is: we just pass the meta-request to the backend, and it can refuse to accept it (with an error message).
> 
> (obviously any error message must post()ed in the master-thread (or properly locked))
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

danomatika commented:
I could split the meta data reading into a separate function. I rolled it into the readheaderfn since the existing implementations are already looping through the file chunks and I figured they might as well poop out meta chunks when they run into them. Having a separate function means we'd need to loop through the file again, but would make the code a bit cleaner due to the separation. For my experiment, I've added some aiff helper functions to abstract the chunk traversal so this wouldn't be a big deal.

danomatika commented:
I think I will change the meta reading into a separate function, otherwise it makes the readheaderfn unnecessarily complex. Being an extra step is offset by better code and the fact that  asking for meta data is an explicit action from the user and not something that happens by default.

danomatika commented:
Latest commit adds custom soundfile file type-specific error codes. This is mainly used to allow for communicating descriptive read header errors when called within a background thread but could be extended for use in other functions.

danomatika commented:
Latest commit adds a -meta flag to `[soundfiler]` write. This allows for up to 8 meta messages to be written before the samples are written. Since the meta messages could be variable length and type, I added a "--" flag which stops flag parsing which allows for the following message:

`[write -aiff -meta title foo -meta artist bar --meta marker 1000 hello\ world -- test.wav array1]`

I also added "--" flags for `[readsf~]` & `[writesf~]` as it also allows for using filenames which start with "-". It's an edge case, but this is a useful convention to have, probably in other objects as well.

danomatika commented:
Latest commits flesh out `[soundfiler]` ascii read/write by adding write support, sample & info output, and fixes a few bugs, ie. crash when reading ascii without tables.

danomatika commented:
Ok, I think I'm happy with this now. I will quickly try writing a libsndfile backend for testing.

Spacechild1 commented:
Cool. I'm currently developing a custom audio streaming protocol with a codec plugin API and I have to write an "opus" plugin in the next few days. I'll take the opportunity and try to write an "opus" plugin for Pd with your soundfile API, just to see if something's missing.

danomatika commented:
I'm sure there are details to iron out and it will be good to try using it first.

danomatika commented:
I managed to get libsndfile reading working as well as error strings, etc. It didn't take much code, really.

There are still issues with the architecture regarding passing in the desired format since it';s not as explicit as a single file format implementation and I also realized there isn't a way to pass in backend-specific flags other than the meta data arguments, which is probably not ideal.

I think I will manually port over relevant changes here back into the main `update/soundfile` branch for Miller to consider without pushing the whole thing yet. I think it makes sense to get some feedback for Christof and adjust things until we are generally satisfied. Also, I simply cannot work on this any more and it already has felt like a never ending "rabbit hole."

danomatika commented:
A few things to think about:

After looking at libsndfile, we could add a t_soundfile_type format int for bit-masked flags. This could be used for endianness, int 16, int 24, and float sample format, etc. It would also be more future-proof in that newer flags could be added without fundamentally changing the API.

Also from libsndfile: should we drop the `readheaderfn` and `writeheaderfn` in favor or simply opening a file?

I'm unsure of how to handle a type backend that doesn't really need `isheaderfn`, as in a multi-format backend such as libsndfile doesn't have a simple magic number or string to check against. This is also a similar issues with `hasextensionfn` and `addextensionfn`.

Following the thought above, perhaps the built-in backends should be integrated into backend wrapper and the backend concept shifted from "single file format" to "multi-file format."

What kind of output lists do we send out when something like the header size or endianness is irrelevant? I'm thinking of possibly using headersize = -1 and endianness = "n" to stand for unknown/native values. I suppose the same is true for bytespersample.

danomatika commented:
I’m going to close this for now. We can return to this later.
